sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,340 @@
+module d.sync.sharedlock;
+
+import d.sync.mutex;
+
+/**
+ * SharedLock allows either one exclusive holder, or N shared holders. While any thread
+ * has the lock shared, one looking to exclusively lock must wait. When
+ * exclusive lock is requested, no new requests to lock will be granted, but
+ * existing holders can unlock.
+ */
+struct SharedLock {
+private:
+	shared Mutex mutex;
+	size_t count;
+	enum size_t Exclusive = 1 << (size_t.sizeof * 8 - 1);
+
+public:
+	/**
+	 * Lock unexclusively.
+	 *
+	 * Can be locked by any number of threads unexclusively. An exclusive
+	 * lock cannot be obtained until all unexclusive locks are released.
+	 */
+	void sharedLock() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).sharedLockImpl();
+	}
+
+	void sharedUnlock() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).exclusiveUnlockImpl();
+	}
+
+	/**
+	 * Lock exclusively.
+	 *
+	 * Upon return, no other threads can hold this locked.
+	 */
+	void exclusiveLock() shared {
+		mutex.lock();
+
+		(cast(SharedLock*) &this).exclusiveLockImpl();
+	}
+
+	void exclusiveUnlock() shared {
+		assert(mutex.isHeld());
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).unlockWriteImpl();
+	}
+
+private:
+	bool noWriteLock() {
+		return count < Exclusive;
+	}
+
+	void sharedLockImpl() {
+		assert(mutex.isHeld());
+		mutex.waitFor(noWriteLock);
+		++count;
+	}
+
+	bool hasExclusiveLock() {
+		return count == Exclusive;
+	}
+
+	void exclusiveLockImpl() {
+		assert(mutex.isHeld());
+		// Wait for no other exclusive lock.
+		mutex.waitFor(noWriteLock);
+		count += Exclusive;
+		mutex.waitFor(hasExclusiveLock);
+	}
+
+	void exclusiveUnlockImpl() {
+		assert(mutex.isHeld());
+		assert((count & ~Exclusive) > 0);
+		--count;
+	}
+
+	void unlockWriteImpl() {
+		assert(mutex.isHeld());
+		assert(count == Exclusive);
+		count = 0;
+	}
+}
+
+unittest sharedLocks {
```

Comment:
You might want to add a comment for each section of this test to explain what procedure is followed.